Log error once in nested long calls

arcusfelis opened the pull request:
The PR reduces the amount of noise reported in "MIM-2073 Multiple errors in the logs on cluster startup with CETS" issue.

Before the same error was reported 3 times from 3 different places (we do over-reporting):

```
when=2023-10-17T19:10:31.083740+00:00 level=error what=long_task_failed reason={{nodedown,'mongooseim@mongooseim-1.mongooseim.default.svc.cluster.local'},{gen_server,call,[<30887.545.0>,other_servers,infinity]}} pid=<0.567.0> at=cets_long:run_tracked/2:63 stacktrace="gen_server:call/3:401 cets_long:run_tracked/2:54 cets_join:get_pids/1:192 cets_join:'-check_fully_connected/2-lc$^0/1-0-'/1:264 cets_join:'-check_fully_connected/2-lc$^0/1-0-'/1:264 cets_join:check_fully_connected/2:264 cets_join:check_pids/4:199 cets_join:join2/4:109" server=<30887.545.0> msg=other_servers pid=<30887.545.0> node=mongooseim@mongooseim-1.mongooseim.default.svc.cluster.local
when=2023-10-17T19:10:31.084280+00:00 level=error what=long_task_failed reason={{nodedown,'mongooseim@mongooseim-1.mongooseim.default.svc.cluster.local'},{gen_server,call,[<30887.545.0>,other_servers,infinity]}} pid=<0.567.0> at=cets_long:run_tracked/2:63 stacktrace="gen_server:call/3:401 cets_long:run_tracked/2:54 cets_join:get_pids/1:192 cets_join:'-check_fully_connected/2-lc$^0/1-0-'/1:264 cets_join:'-check_fully_connected/2-lc$^0/1-0-'/1:264 cets_join:check_fully_connected/2:264 cets_join:check_pids/4:199 cets_join:join2/4:109" local_pid=<0.545.0> remote_pid=<30886.545.0> remote_node=mongooseim@mongooseim-0.mongooseim.default.svc.cluster.local table=cets_cluster_id
when=2023-10-17T19:10:31.086471+00:00 level=error what=long_task_failed reason={{nodedown,'mongooseim@mongooseim-1.mongooseim.default.svc.cluster.local'},{gen_server,call,[<30887.545.0>,other_servers,infinity]}} pid=<0.561.0> at=cets_long:run_tracked/2:63 stacktrace="gen_server:call/3:401 cets_long:run_tracked/2:54 cets_join:get_pids/1:192 cets_join:'-check_fully_connected/2-lc$^0/1-0-'/1:264 cets_join:'-check_fully_connected/2-lc$^0/1-0-'/1:264 cets_join:check_fully_connected/2:264 cets_join:check_pids/4:199 cets_join:join2/4:109" long_task_name=join local_pid=<0.545.0> remote_pid=<30886.545.0> remote_node=mongooseim@mongooseim-0.mongooseim.default.svc.cluster.local table=cets_cluster_id
```

Once inside `cets` call and 2 times inside the `cets_join` module.

Changes:
- This PR removes 2 extra logs.
- Only the most nested call would be logged. In our case - we log from `cets` module. `cets_join` would remain silent.
- We add `task_failed` wrapping when we re-throw the error, to avoid it being reported by the nested long (call from `cets_long` module) call.
- This PR only removes log duplication, in the real life you would still see a lot of messages: because of retry logic and because a message could be potentially logged for each unique table name. At least the amount of logged messages is less now.
- Updated the version of the debug log handler dependency. Because OTP Logger has real hard time when add_handler/remove_handler are called from the different processes (there are race condition bugs there, we would need to create a bug report to the OTP. For now, we would have to use `global:transaction` when adding log handlers in our tests).

arcusfelis committed:
Add caller_pid to check pids

Useful for distinguising reports from the parallel tasks for example
Sadly, MongooseIM does not report pid in logs

arcusfelis committed:
Add a test prototype

arcusfelis committed:
Add meck, simulate error

arcusfelis committed:
Log only one message instead of three when failing to join the cluster

Log only first time when having nested long tasks
Remove long from the task tags to not to confuse with actually long tasks
(though, progress logging would still say "long")

arcusfelis committed:
Properly clean in logging_when_failing_join_with_disco

arcusfelis committed:
Add long_call_fails_because_linked_process_dies to improve coverage

And to check that logging actually works there

chrzaszcz commented on src/cets_long.erl:
```diff
@@ -48,28 +48,30 @@ run_spawn(Info, F) ->
         Fun()
     catch
-        Class:Reason:Stacktrace ->
+        Class:Reason:Stacktrace when element(1, Reason) =/= task_failed ->
```
Interesting, so this guard can fail but the failure is just treated as non-matching.

arcusfelis replied:
true, fixed.
So, I also changed format of assertException in some tests, after that.

chrzaszcz reviewed (approved):
Looks 👌

arcusfelis committed:
Add tests for nested call logging

arcusfelis committed:
Fixed race conditions

We use the latest version of debug_log_handler, which has fixes
for race conditions in the Logger

arcusfelis committed:
Use logger_debug_h 0.2.0 from Hex

chrzaszcz reviewed (approved):
Looks good :+1:
